From 3d6676c762c676e6300c6f90df12f18cc69a4612 Mon Sep 17 00:00:00 2001
From: Priya N S <priya.ns@ti.com>
Date: Mon, 6 Dec 2021 15:06:05 +0530
Subject: [PATCH 2/2] arm: mach-omap2: load remoteproc IPU1/IPU2 from SPI

Update the spl_boot_ipu() API to load IPU1/IPU2 remoteproc firmware
from the SPI flash incase of SPI/QPSI booting.

Signed-off-by: Priya N S <priya.ns@ti.com>
---
 arch/arm/mach-omap2/boot-common.c | 41 +++++++++++++++++++++++++------
 include/configs/am57xx_evm.h      | 14 +++++++++++
 2 files changed, 47 insertions(+), 8 deletions(-)

diff --git a/arch/arm/mach-omap2/boot-common.c b/arch/arm/mach-omap2/boot-common.c
index f264e3ec91..f537acd575 100644
--- a/arch/arm/mach-omap2/boot-common.c
+++ b/arch/arm/mach-omap2/boot-common.c
@@ -228,10 +228,22 @@ void spl_boot_ipu(void)
 	    !IS_ENABLED(CONFIG_REMOTEPROC_TI_IPU))
 		return;
 
-	size = load_firmware("dra7-ipu1-fw.xem4", &loadaddr);
-	if (size <= 0) {
-		pr_err("Firmware loading failed\n");
-		goto skip_ipu1;
+	if(BOOT_DEVICE_SPI == spl_boot_device())
+	{
+		ret = spl_spi_load_firmware("dra7-ipu1-fw.xem4", &loadaddr, CONFIGS_SYS_SPI_IPU1_OFFS);
+		if(ret)
+		{
+			debug("%s: Firmware loading failed for ipu1\n", __func__);
+			goto skip_ipu1;
+		}
+	}
+	else
+	{
+		size = load_firmware("dra7-ipu1-fw.xem4", &loadaddr);
+		if (size <= 0) {
+			debug("%s: Firmware loading failed for ipu1\n", __func__);
+			goto skip_ipu1;
+		}
 	}
 
 	enable_ipu1_clocks();
@@ -257,10 +269,23 @@ void spl_boot_ipu(void)
 
 skip_ipu1:
 	loadaddr = IPU2_LOAD_ADDR;
-	size = load_firmware("dra7-ipu2-fw.xem4", &loadaddr);
-	if (size <= 0) {
-		pr_err("Firmware loading failed for ipu2\n");
-		return;
+
+	if(BOOT_DEVICE_SPI == spl_boot_device())
+	{
+		ret = spl_spi_load_firmware("dra7-ipu2-fw.xem4", &loadaddr, CONFIGS_SYS_SPI_IPU2_OFFS);
+		if(ret)
+		{
+			debug("%s: Firmware loading failed for ipu2\n", __func__);
+			return;
+		}
+	}
+	else
+	{
+		size = load_firmware("dra7-ipu2-fw.xem4", &loadaddr);
+		if (size <= 0) {
+			debug("%s: Firmware loading failed for ipu2\n", __func__);
+			return;
+		}
 	}
 
 	enable_ipu2_clocks();
diff --git a/include/configs/am57xx_evm.h b/include/configs/am57xx_evm.h
index c47ffccff1..d9361415e7 100644
--- a/include/configs/am57xx_evm.h
+++ b/include/configs/am57xx_evm.h
@@ -78,6 +78,20 @@
 #define CONFIG_SYS_SPI_ARGS_OFFS        0x140000
 #define CONFIG_SYS_SPI_ARGS_SIZE        0x80000
 
+#define CONFIGS_SYS_SPI_KERNEL_SIZE      0x800000
+#define CONFIGS_SYS_SPI_LOGO_OFFS       (CONFIGS_SYS_SPI_KERNEL_SIZE +\
+                                        (0x1E0000))
+#define CONFIGS_SYS_SPI_LOGO_SIZE        0x80000
+#define CONFIGS_SYS_SPI_RPROC_BIN_OFFS  (CONFIGS_SYS_SPI_LOGO_OFFS +\
+                                        CONFIGS_SYS_SPI_LOGO_SIZE)
+
+#define CONFIGS_SYS_SPI_RPROC_BIN_SIZE   0x1000000
+#define CONFIGS_SYS_SPI_IPU1_BIN_SIZE    0x800000
+#define CONFIGS_SYS_SPI_IPU2_BIN_SIZE    0x800000
+#define CONFIGS_SYS_SPI_IPU1_OFFS       (CONFIGS_SYS_SPI_RPROC_BIN_OFFS)
+#define CONFIGS_SYS_SPI_IPU2_OFFS       (CONFIGS_SYS_SPI_IPU1_OFFS +\
+        CONFIGS_SYS_SPI_IPU1_BIN_SIZE)
+
 /* SPI SPL */
 
 #endif /* __CONFIG_AM57XX_EVM_H */
-- 
2.17.0

